packages/experimental-identity-and-auth/src/memoizeIdentityProvider.ts

```diff
@@ -0,0 +1,87 @@
+import { Identity, IdentityProvider } from "@smithy/types";
+
+/**
+ * @internal
+ * This may need to be configurable in the future, but for now it is defaulted to 5min.
+ */
+export const EXPIRATION_MS = 300_000;
+
+/**
+ * @internal
+ */
+export const isIdentityExpired = (identity: Identity) =>
+  doesIdentityRequireRefresh(identity) && identity.expiration!.getTime() - Date.now() < EXPIRATION_MS;
+
+/**
+ * @internal
+ */
+export const doesIdentityRequireRefresh = (identity: Identity) => identity.expiration !== undefined;
+
+/**
+ * @internal
+ */
+export interface MemoizedIdentityProvider<IdentityT extends Identity> {
+  (options?: Record<string, any> & { forceRefresh?: boolean }): Promise<IdentityT>;
+}
+
+/**
+ * @internal
+ */
+export const memoizeIdentityProvider = <IdentityT extends Identity>(
```

Comment:
This is based on `memoize()` in `@smithy/property-provider`, except it also handles `undefined` values.